### ISSUE TITLE
[Windows] Prevent sync trigger when file has changed but has state open (by another app) (locked file)

### ISSUE THREAD

**Issue by mmattel** [owncloud/client#6822]
### Expected behaviour
Client should not start trying syncing a file if the file is in an open state.
The open state hapens if you eg work on an Excel or Word document.
Changes are written to that file on every save, but the file is locked by the app.

### Actual behaviour
The file as described in the example above triggers on any change a sync but it is in an open state by the app. This conflicts and you see a red crossed button on the file sync status as long the file is open. When the file gets released by the app, syncing is fine and things turn green.

### Steps to reproduce
1. Create or open an Office document, save it, but keep it open
2. Watch the client trying to sync
3. Watch the filesytem with regards to the file sync status
4. Close the Office document and see 2) and 3)

### Server configuration
Operating system: Ubuntu 16.04

Web server: nginx

Database: mysql

PHP version: 7.0

ownCloud version: 10.0.10 (stable)

Storage backend (external storage): smb, ftp, dropbox, local

### Client configuration
Client version: 2.5.0

Operating system: W10x64

OS language: DE

**Comment by michaelstingl** [comment]
It seems this really difficult in the Win/Office world. Here is an overview from an user in the heise.de forum:
https://www.heise.de/forum/iX/News-Kommentare/ownCloud-Neue-Synchronisationsfunktion-fuer-den-Desktop/Re-Ich-wuerde-Filelocking-im-Desktop-Client-dringender-benoetigen/posting-32458876/show/

**Comment by mmattel** [comment]
The point is, that the friends at MS create a new file containing an ``~`` sign in the filename (somewhere in the filename, see link above).
BUT MS also locks the original file and prevents double writing on. When the application closed things get cleand up and you have a correct freed file with new content.
My request is to not only check for state ``has changed`` but also for state ``is write opened`` by another app (no care on ``is read opened``). And do not sync in case both are present, only in first case.

**Comment by mmattel** [comment]
Another usecase which is much more highlighting this problematic is iTunes.
If you have your library in a folder which gets synced with the client, you see this much more clearly. 
iTunes has two files which trigger this issue: the database ``iTunes Library.itl`` and a file called ``sentinel``.
iTunes does many changes in the database while the app is open. ``sentinel`` gets deleted when iTunes is closed. Both get sync errors constantly during iTunes is open.

**Comment by mmattel** [comment]
More impact (really bad)
- I opened a already synced Word document, did a change, Word still open, client tries to sync - fails.
- Closing word, file is now ok, but before sync starts, I moved the file out of the sync folder.
- The file is now in the new (unsynced) folder, sync starts - and -
it gets downloaded again... - now I have two of them - one in the sync folder and one in the non sync folder 🤦‍♂️

**Comment by guruz** [comment]
@ckamm How does this relate to https://github.com/owncloud/client/blob/master/ChangeLog#L392 https://github.com/owncloud/client/blob/master/ChangeLog#L379 ?

@mmattel Which office version btw?

@mmattel Are you talking about syncing down server side changes or syncing up client side changes?

**Comment by mmattel** [comment]
Office 365, v1809
Windows - client - core - SMB backend
Changes are always done at the client. Nobody else accessed the backend (no concurrent access)
Anything gets triggered initiated from the client side, I do not see what the negotioation upwards is, only the issues...

**Comment by ckamm** [comment]
@guruz When the client notices the file is locked it'll keep track of it and check whether it becomes unlocked in the future (checks every 20s). Once it is, it schedules the folder for a sync run again.

This should not produce the described behavior: the follow-up sync should just notice the file is gone locally and remove it on the remote. I can try to reproduce it.

**Comment by guruz** [comment]
OK I think I understand the issue now... but...
Isn't it bad if we don't sync a Word file just because it's open? 
Shouldn't we only get errors in the case there was a server-side change too? I thought locking is about preventing writing, not about preventing read+upload

(CC @ogoffart )

**Comment by mmattel** [comment]
For clarification. Syncing a file which is open read only is totally fine, but syncing a file beeing open for writing is bad...

**Comment by ckamm** [comment]
Indeed, there are different kinds of locks. Some prevent writing, some reading+writing (actually it's way more complicated) and the lock file watcher can be fed from propagatedownload (renameReplace fail) as well as from propagateupload (file open fail). 

I was thinking of "if we get a file-touched notification for something that's already in the lock-watcher, don't trigger a sync run". That should avoid the client trying to upload a file over and over again that's currently being worked on by the user in excel while preserving the initial attempt and error.

**Comment by ckamm** [comment]
@mmattel I couldn't reproduce the problem where you moved the locked file away and it was re-downloaded. Could you send me a log file for it?

**Comment by mmattel** [comment]
sure, will do today

**Comment by ckamm** [comment]
I also found a problem while digging into that code, but I don't see how it'd would cause the particular symptom you described.

**Comment by mmattel** [comment]
@ckamm this drives me nuts ! :neckbeard:
2 Windows machines, nothing changed. Yesterday when I was in rush and needed to support my son quickly, it happened on the itunes .itl file as described and as expected, also on word.
Today when I had time and did testing, the behaviour did NOT occur.
AHHHH... (:rage3: censored,) in these cases you feel like an i*** :monkey:

**Comment by ogoffart** [comment]
@mmattel  Are you saying that the bug was not fixed by the patch? (which is in the 2.5.1 nightly)

**Comment by guruz** [comment]
This should be fixed in 2.5.2 rc2.
Please try it out: https://download.owncloud.com/desktop/testing/ownCloud-2.5.2.11335.11186-rc2.en-US.msi
And re-open the issue if needed. (and if it works also comment if you want..)

**Comment by mmattel** [comment]
@jnweiger I am currently not able to test because I am for the next two weeks on the road and do not have access to my Windows test environment.

**Comment by lazawan** [comment]
@mmattel could you retest this issue, your feedback is important  :)

**Comment by mmattel** [comment]
Running 2.5.4 stable on Windows 10x64.
Tested with open
- Word document and editing/saving
- iTunes where the database is in a synced folder

Watching during changes sync activities - no errors :smile:

**Comment by mmattel** [comment]
Needs maybe a test on macOS...
